### PR TITLE
Optimisation of Management Folder.

### PR DESCRIPTION
**Pull request by Ksteeds**

### CODE REVIEW

File: macrosynergy/management/check_availability.py [hunk 1]
```diff
@@ -3,130 +3,136 @@
 import matplotlib.pyplot as plt
 import seaborn as sns
 from typing import List, Union, Tuple
+import time
 
-from macrosynergy.management.simulate_quantamental_data import make_qdf
-from macrosynergy.management.shape_dfs import reduce_df
+from Simulate_Quantamental import make_qdf
```
Comment (lsimonsen) [inline macrosynergy/management/check_availability.py:8]: The line that is failing + lower cases for modules (files) ie. `simulate_quantamental`

**Comment by lsimonsen** [comment]
A general comment on style, we should try to adhere to [PEP8](https://www.python.org/dev/peps/pep-0008/).

One thing that is worth doing is in the PyCharm editor to set the linewidth to 79 characters instead ofg 150 (or what it currently is).

Also for the comments in general, learning a bit of Markdown will get you a long way (especially for REDME.md and GitHub).

**Comment by lsimonsen** [comment]
@Ksteeds you will need to update branch to the latest version relative to develop as well. You can do this on GitHub directly by pressing the *Update branch* bottom.

File: macrosynergy/management/shape_dfs.py [hunk 2]
```diff
@@ -37,154 +43,143 @@ def reduce_df(df: pd.DataFrame, xcats: List[str] = None,  cids: List[str] = None
...
             dfx = dfx[~(filt1 & filt2 & filt3)]
-
-    xcats_in_df = dfx['xcat'].unique()
-    if xcats is None:
-        xcats = sorted(xcats_in_df)
-    else:
-        missing = sorted(set(xcats) - set(xcats_in_df))
-        if len(missing) > 0:
-            print(f'Missing cross sections: {missing}')
-            xcats.remove(missing)
```
Comment (lsimonsen) [inline macrosynergy/management/shape_dfs.py:48]: The printing of the information that we are missing categories are an important part of the function.

File: macrosynergy/management/shape_dfs.py [hunk 3]
```diff
@@ -37,154 +43,143 @@ def reduce_df(df: pd.DataFrame, xcats: List[str] = None,  cids: List[str] = None
...
             cids_in_df = [cid for cid in df_uns[i] if cid in cids_in_df]
-    else:
-        cids_in_df = dfx['cid'].unique()
+    
+    cids = data_frame(cids_in_df, cids, 'cids')
+    dfx = dfx[dfx['cid'].isin(cids)]
 
-    if cids is None:
-        cids = sorted(cids_in_df)
+    ## Would there ever be any duplicates ?
```
Comment (lsimonsen) [inline macrosynergy/management/shape_dfs.py:60]: Theoretically no, however, never trust data, and always better for data in the wild to double check (it isn't that computational expensive, relative to the errors in the output you can get).

File: macrosynergy/management/shape_dfs.py [hunk 4]
```diff
@@ -37,154 +43,143 @@ def reduce_df(df: pd.DataFrame, xcats: List[str] = None,  cids: List[str] = None
...
+    if out_all: return dfx, xcats, cids
     else:
-        if not isinstance(cids, list):
-           cids = [cids]
-        missing = sorted(set(cids) - set(cids_in_df))
-        if len(missing) > 0:
-            print(f'Missing cross sections: {missing}')
-        cids = sorted(list(set(cids).intersection(set(cids_in_df))))
-        dfx = dfx[dfx['cid'].isin(cids)]
```
[the remark this answers is not in the document]
Comment (lsimonsen) [reply macrosynergy/management/shape_dfs.py:69]: Similar to above, the pring of the missing cross sections is a key bit.

File: macrosynergy/management/shape_dfs.py [hunk 5]
```diff
@@ -37,154 +43,143 @@ def reduce_df(df: pd.DataFrame, xcats: List[str] = None,  cids: List[str] = None
...
-    :param <int> years: Number of years over which data are aggregated. Supersedes freq and does not allow lags,
-        Default is None, i.e. no multi-year aggregation.
-    :param <str> val: name of column that contains the values of interest. Default is 'value'.
-    :param <str> freq: letter denoting frequency at which the series are to be sampled.
-        This must be one of 'D', 'W', 'M', 'Q', 'A'. Default is 'M'.
-    :param <int> lag: Lag (delay of arrival) of second category in periods as set by freq. Default is 0.
-        Note: for analyses with dependent and explanatory categories, the second takes the role of the explanatory.
-    :param <int> fwin: Forward moving average window of first category. Default is 1, i.e no average.
-        Note: this parameter is used mainly for target returns as dependent variables.
-    :param <List[str]> xcat_aggs: Exactly two aggregation methods. Default is 'mean' for both.
```
Comment (lsimonsen) [inline macrosynergy/management/shape_dfs.py:100]: Re-instate docstring.

File: macrosynergy/management/shape_dfs.py [hunk 6]
```diff
@@ -37,154 +43,143 @@ def reduce_df(df: pd.DataFrame, xcats: List[str] = None,  cids: List[str] = None
...
 
-    """
+## Create two custom category dataframes with appropriate frequency and lags suitable for analysis.
+## The subroutine will receive a dataframe that has been reduced to the two respective macroeconomic indicators whose relationship is to be analysed over the respective countries.
+def categories_df(df, xcats, fields_cats = [], fields_cids = [], cids_cats = {}, cids = [], val = 'value',
+                  start = None, end = None, blacklist = None, years = None, freq = 'M', lag = 0,
+                  fwin = 1, xcat_aggs = ('mean', 'mean')):
 
     assert freq in ['D', 'W', 'M', 'Q', 'A']
-    assert not (years is not None) & (lag != 0), 'Lags cannot be applied to year groups'
```
Comment (lsimonsen) [inline macrosynergy/management/shape_dfs.py:107]: Why remove?

File: macrosynergy/management/simulate_quantamental_data.py [hunk 7]
```diff
@@ -4,109 +4,201 @@
...
-    """Create an autocorrelated dataseries as numpy array
+    ar_params = np.r_[1, -ar_coef]
+    ## AR(1) with coefficient 0.75 - high persistance from the first period.
+    ar_proc = ArmaProcess(ar_params)
+    ser = ar_proc.generate_sample(no_days)
 
-    :param <int> nobs: number of observations.
-    :param <float> mean: mean of values, default is zero.
-    :param <float> sd_mult: standard deviation multipliers of values, default is 1. This affects non-zero means.
-    :param <float> ar_coef: autoregression coefficient (between 0 and 1): default is 0.75.
```
Comment (lsimonsen) [inline macrosynergy/management/simulate_quantamental_data.py:15]: Re-instate docstring.

File: macrosynergy/management/simulate_quantamental_data.py [hunk 8]
```diff
@@ -4,109 +4,201 @@
...
-        'mean_add': float of country-specific addition to any category's mean;
-        'sd_mult': float of country-specific multiplier of an category's standard deviation.
-    :param <pd.DataFrame> df_xcats: dataframe with parameters by xcat. Row indices are cross-sections. Columns are:
-        'earliest': string of earliest date (ISO) for which category values are available;
-        'latest': string of latest date (ISO) for which category values are available;
-        'mean_add': float of category-specific addition;
-        'sd_mult': float of country-specific multiplier of an category's standard deviation;
-        'ar_coef': float between 0 and 1 denoting set autocorrelation of the category;
-        'back_coef': float, coefficient with which communal (mean 0, SD 1) background factor is added to categoy values.
-    :param <float> back_ar: float between 0 and 1 denoting set autocorrelation of the background factor.
```
Comment (lsimonsen) [inline macrosynergy/management/simulate_quantamental_data.py:43]: Again keep docstring

**Review by lsimonsen — CHANGES_REQUESTED** [review]

**Review by rsueppel — APPROVED** [review]

**Pull request closed without merging**